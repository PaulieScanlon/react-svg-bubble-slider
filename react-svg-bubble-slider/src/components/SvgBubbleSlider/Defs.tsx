import React, { FunctionComponent, memo } from 'react'

export const Defs: FunctionComponent = memo(() => (
  <defs>
    <filter id="goo" colorInterpolationFilters="sRGB">
      <feGaussianBlur in="SourceGraphic" stdDeviation="8" result="blur" />
      <feColorMatrix
        in="blur"
        mode="matrix"
        values="1 0 0 0 0  0 1 0 0 0  0 0 1 0 0  0 0 0 31 -12"
        result="cm"
      />
    </filter>
  </defs>
))

Defs.displayName = 'Defs'
