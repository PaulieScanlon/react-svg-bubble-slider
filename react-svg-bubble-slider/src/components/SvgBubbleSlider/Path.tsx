import React, { FunctionComponent, memo } from 'react'
import { createIconPathsMarkup } from '../utils'

interface PathProps {
  /** Array of paths from iconPaths */
  paths: string[]
}

export const Path: FunctionComponent<PathProps> = memo(({ paths }) => {
  return <g dangerouslySetInnerHTML={createIconPathsMarkup(paths)} />
})
