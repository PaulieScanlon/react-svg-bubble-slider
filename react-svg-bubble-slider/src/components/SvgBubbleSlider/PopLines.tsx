import React, { FunctionComponent, RefObject, useRef, useEffect } from 'react'

import { gsap } from 'gsap'
import { DrawSVGPlugin } from 'gsap/DrawSVGPlugin'
gsap.registerPlugin(DrawSVGPlugin)
interface PopLinesProps {
  /** The color of the dots, speech bubble background and speech bubble text and pop lines */
  primaryColor?: string
  /** Is the animation happening */
  animationState: boolean
}

const lineAttributes = [
  {
    x1: '107.923',
    y1: '21.37',
    x2: '116.935',
    y2: '4',
  },
  {
    x1: '59.335',
    y1: '24.909',
    x2: '50.5',
    y2: '8.057',
  },
  {
    x1: '21.789',
    y1: '56.11',
    x2: '4',
    y2: '53.67',
  },
  {
    x1: '26.782',
    y1: '98.86',
    x2: '8.057',
    y2: '108.34',
  },
  {
    x1: '65.885',
    y1: '125.86',
    x2: '56.43',
    y2: '145.37',
  },
  {
    x1: '112.429',
    y1: '121.342',
    x2: '121.03',
    y2: '142.564',
  },
  {
    x1: '147.535',
    y1: '93.82',
    x2: '168.155',
    y2: '101.45',
  },
  {
    x1: '149.742',
    y1: '49.01',
    x2: '168.155',
    y2: '42.56',
  },
]

export const PopLines: FunctionComponent<PopLinesProps> = ({
  primaryColor,
  animationState,
}: PopLinesProps) => {
  const popLinesRef = useRef(null)

  const lineRefs: RefObject<SVGPathElement>[] = []

  useEffect(() => {
    lineRefs.map((_, index: number) => {
      animationState
        ? gsap.to(lineRefs[index], 0.2, {
            scale: 2.2,
            drawSVG: '100% 100%',
            ease: 'linear',
          })
        : gsap.set(lineRefs[index], {
            transformOrigin: '50% 50%',
            drawSVG: '0% 100%',
            scale: 0,
          })
    })
  }, [animationState])

  useEffect(() => {
    gsap.set(popLinesRef.current, {
      visibility: 'visible',
    })
  }, [])

  return (
    <g
      ref={popLinesRef as RefObject<any>}
      className="pop-lines"
      fill="none"
      stroke={primaryColor}
      strokeWidth="4"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeMiterlimit="10"
      transform={`matrix(0.8,0,0,0.8,125,0)`}
      style={{
        visibility: 'hidden',
      }}
    >
      {lineAttributes.map((line, index: number) => (
        <line ref={(ref) => lineRefs.push(ref as any)} key={index} {...line} />
      ))}
    </g>
  )
}

PopLines.displayName = 'PopLines'
