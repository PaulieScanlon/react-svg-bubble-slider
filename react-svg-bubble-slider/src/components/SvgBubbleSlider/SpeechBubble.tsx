import React, {
  FunctionComponent,
  RefObject,
  useRef,
  useEffect,
  useState,
} from 'react'

import { gsap } from 'gsap'

import { SpeechBubbleProps } from './types'

const START_Y = 60

export const SpeechBubble: FunctionComponent<SpeechBubbleProps> = ({
  primaryColor,
  secondaryColor,
  currentReaction,
}: SpeechBubbleProps) => {
  const speechBubblesRef = useRef(null)
  const [tl] = useState({
    timeline: gsap.timeline(),
  })

  useEffect(() => {
    tl.timeline
      .set(speechBubblesRef.current, {
        rotation: 45,
        visibility: 'hidden',
        scale: 0,
        y: START_Y,
        transformOrigin: `50% 100%`,
      })
      .to(speechBubblesRef.current, 0.9, {
        rotation: 0,
        visibility: 'visible',
        ease: 'elastic(1, 0.6)',
        scale: 0.9,
        y: 40,
      })
      .to(
        speechBubblesRef.current,
        0.6,
        {
          ease: 'easltic(1, 0.6)',
          scale: 1,
        },
        '-=1'
      )
  }, [])

  useEffect(() => {
    currentReaction.length > 0
      ? tl.timeline.play()
      : tl.timeline.pause() && tl.timeline.seek(0)
  }, [currentReaction])

  return (
    <g
      ref={speechBubblesRef as RefObject<any>}
      style={{
        pointerEvents: 'none',
        visibility: 'hidden',
      }}
      transform={`matrix(0.8,0,0,0.8,91.5,${START_Y})`}
    >
      <path
        className="speech-bubble-stroke"
        stroke={primaryColor}
        strokeWidth="14"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeMiterlimit="10"
        d="
		M69.361,112.063l16.377,28.99l16.562-28.94c37.277-5.125,65.195-27.13,65.195-53.494C167.496,28.456,130.766,4,85.515,4
		C40.275,4,4,28.456,4,58.619c0,26.291,27.361,48.249,65.361,53.448V112.063z"
      ></path>
      <path
        className="speech-bubble-fill"
        d="M69.361,109.063l16.377,28.99l16.562-28.94
		c37.277-5.125,65.195-27.13,65.195-53.494C167.496,25.456,130.766,1,85.515,1C40.275,1,4,25.456,4,55.619
    c0,26.291,27.361,48.249,65.361,53.448V109.063z"
        fill={secondaryColor}
      ></path>
      <text
        className="speech-bubble-text"
        x="85"
        y="67"
        fill={primaryColor}
        style={{
          textAnchor: 'middle',
        }}
      >
        {currentReaction}
      </text>
    </g>
  )
}

SpeechBubble.displayName = 'SpeechBubble'
