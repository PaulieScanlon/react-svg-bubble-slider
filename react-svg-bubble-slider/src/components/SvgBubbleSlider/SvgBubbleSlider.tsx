import React, { Fragment, FunctionComponent, useState } from 'react'

import { SvgBubbleSliderProps } from './types'
import theme from '../../theme'

import { Timeline } from './Timeline'

export const SvgBubbleSlider: FunctionComponent<SvgBubbleSliderProps> = ({
  children,
  primaryColor = theme.colors.primary,
  secondaryColor = theme.colors.background,
  icons,
  showSpeechBubble,
}: SvgBubbleSliderProps) => {
  const [currentReaction, setCurrentReaction] = useState<any>('')

  return (
    <Fragment>
      <div
        className="react-svg-bubble-slider"
        style={{
          display: 'flex',
          flexDirection: 'column',
          justifyContent: 'center',
          position: 'relative',
          width: '100%',
        }}
      >
        <Timeline
          onAnimationComplete={(reaction) => setCurrentReaction(reaction)}
          primaryColor={primaryColor}
          secondaryColor={secondaryColor}
          icons={icons}
          showSpeechBubble={showSpeechBubble}
        />
      </div>
      <div className="svg-bubble-action">
        {children &&
          children({
            reaction: currentReaction,
          })}
      </div>
    </Fragment>
  )
}
