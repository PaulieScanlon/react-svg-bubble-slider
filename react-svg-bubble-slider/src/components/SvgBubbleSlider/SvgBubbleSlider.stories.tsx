import React from 'react'
import { action } from '@storybook/addon-actions'

import { ReactionType } from '../SvgBubbleSlider/types'
import { SvgBubbleSlider } from '.'

import { Box } from 'theme-ui'

const handleReaction = (reaction: ReactionType) => {
  action('handleReaction')(reaction)
}

export default {
  title: 'Props',
  parameters: {
    component: SvgBubbleSlider,
    componentSubtitle:
      'Have a browse of the props to see how you can customise SvgBubbleSlider. You can see the usage by clicking on "Show code" in the bottom right of the preview panel ðŸ‘‡',
  },
}

export const usage = () => <SvgBubbleSlider />

export const icons = () => (
  <SvgBubbleSlider icons={['angry', 'sad', 'smile', 'happy']} />
)

icons.story = {
  parameters: {
    docs: {
      storyDescription:
        'Use the `icons` prop to provide a restricted list of icons to display',
    },
  },
}

export const reaction = () => (
  <div style={{ position: 'relative', top: 38 }}>
    <SvgBubbleSlider>
      {({ reaction }: any) => (
        <div style={{ textAlign: 'center', height: 44 }}>
          {reaction && (
            <button onClick={() => handleReaction(reaction)}>{reaction}</button>
          )}
        </div>
      )}
    </SvgBubbleSlider>
  </div>
)

reaction.story = {
  parameters: {
    docs: {
      storyDescription:
        'You can access the current `reaction` via the render prop',
    },
  },
}

export const primaryColor = () => <SvgBubbleSlider primaryColor="#4fc3f7" />

primaryColor.story = {
  parameters: {
    docs: {
      storyDescription:
        'The `primaryColor` prop can be used to change the color of the dots, speech bubble stroke, speech bubble text and pop lines',
    },
  },
}

export const secondaryColor = () => <SvgBubbleSlider secondaryColor="#ffebee" />

secondaryColor.story = {
  parameters: {
    docs: {
      storyDescription:
        'The `secondaryColor` prop can be used to change the color of the reaction icons and the speech bubble fill',
    },
  },
}

export const ThemeUI = () => (
  <Box
    sx={{
      '.speech-bubble-stroke': {
        stroke: 'muted',
      },
      '.speech-bubble-fill': {
        fill: 'background',
      },
      '.speech-bubble-text': {
        fill: 'muted',
        fontSize: '24px',
        textTransform: 'capitalize',
      },
      '.speech-bubble-pop-line': {
        stroke: 'muted',
      },
      '.reaction-icon': {
        fill: 'background',
      },
      '.reaction-dot': {
        fill: 'muted',
      },
      '.svg-bubble-action': {
        minHeight: 2,
      },
    }}
  >
    <SvgBubbleSlider />
  </Box>
)

ThemeUI.story = {
  parameters: {
    docs: {
      storyDescription:
        "If you're using Theme UI modify styles by referencing them by class name via the sx prop on a Theme UI enabled element",
    },
  },
}
