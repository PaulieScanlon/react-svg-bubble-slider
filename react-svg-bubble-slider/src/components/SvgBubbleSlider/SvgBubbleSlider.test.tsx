import { render } from '@testing-library/react'

import { usage } from '../../_docs/SvgBubbleSlider.stories'

// https://github.com/jsdom/jsdom/issues/1664
describe('<SvgBubbleSlider />', () => {
  beforeEach(() => {
    // @ts-ignore
    window.SVGElement.prototype.getBBox = () => ({
      x: 0,
      y: 0,
    })
  })

  afterEach(() => {
    // @ts-ignore
    delete window.SVGElement.prototype.getBBox
  })

  test('renders the component', () => {
    const { asFragment } = render(usage())
    expect(asFragment()).toMatchSnapshot()
  })
})
