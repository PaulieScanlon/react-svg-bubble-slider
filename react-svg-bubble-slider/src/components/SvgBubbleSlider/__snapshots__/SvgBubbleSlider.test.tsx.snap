// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SvgBubbleSlider /> renders the component 1`] = `
<DocumentFragment>
  <div
    class="react-svg-bubble-slider"
    style="display: flex; flex-direction: column; justify-content: center; position: relative; width: 100%;"
  >
    <div
      aria-valuemax="13"
      aria-valuemin="0"
      aria-valuenow="0"
      aria-valuetext=""
      class="svg-timeline"
      role="slider"
      style="align-items: center; display: flex; justify-content: center; margin: auto; position: relative; overflow: hidden; width: 100%;"
      tabindex="0"
    >
      <div
        style="display: flex; margin: auto;"
      >
        <svg
          height="290"
          style="min-width: 280px;"
          viewBox="0,0, 808,290"
          width="808"
        >
          <defs>
            <filter
              color-interpolation-filters="sRGB"
              id="goo"
            >
              <fegaussianblur
                in="SourceGraphic"
                result="blur"
                stdDeviation="8"
              />
              <fecolormatrix
                in="blur"
                mode="matrix"
                result="cm"
                values="1 0 0 0 0  0 1 0 0 0  0 0 1 0 0  0 0 0 31 -12"
              />
            </filter>
          </defs>
          <g
            fill="none"
            stroke="#FF69B4"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-miterlimit="10"
            stroke-width="4"
            style="visibility: visible;"
            transform="matrix(1,0,0,1,316,15)"
          >
            <line
              class="speech-bubble-pop-line"
              style="transform-origin: 50% 50%; transform: scale(0, 0) ;"
              x1="107.923"
              x2="116.935"
              y1="21.37"
              y2="4"
            />
            <line
              class="speech-bubble-pop-line"
              style="transform-origin: 50% 50%; transform: scale(0, 0) ;"
              x1="59.335"
              x2="50.5"
              y1="24.909"
              y2="8.057"
            />
            <line
              class="speech-bubble-pop-line"
              style="transform-origin: 50% 50%; transform: scale(0, 0) ;"
              x1="21.789"
              x2="4"
              y1="56.11"
              y2="53.67"
            />
            <line
              class="speech-bubble-pop-line"
              style="transform-origin: 50% 50%; transform: scale(0, 0) ;"
              x1="26.782"
              x2="8.057"
              y1="98.86"
              y2="108.34"
            />
            <line
              class="speech-bubble-pop-line"
              style="transform-origin: 50% 50%; transform: scale(0, 0) ;"
              x1="65.885"
              x2="56.43"
              y1="125.86"
              y2="145.37"
            />
            <line
              class="speech-bubble-pop-line"
              style="transform-origin: 50% 50%; transform: scale(0, 0) ;"
              x1="112.429"
              x2="121.03"
              y1="121.342"
              y2="142.564"
            />
            <line
              class="speech-bubble-pop-line"
              style="transform-origin: 50% 50%; transform: scale(0, 0) ;"
              x1="147.535"
              x2="168.155"
              y1="93.82"
              y2="101.45"
            />
            <line
              class="speech-bubble-pop-line"
              style="transform-origin: 50% 50%; transform: scale(0, 0) ;"
              x1="149.742"
              x2="168.155"
              y1="49.01"
              y2="42.56"
            />
          </g>
          <g
            style="pointer-events: none; visibility: hidden; transform: translate(0, 0);"
            transform="matrix(0.8,0,0,0.8,336,80)"
          >
            <path
              class="speech-bubble-stroke"
              d="
		M69.361,112.063l16.377,28.99l16.562-28.94c37.277-5.125,65.195-27.13,65.195-53.494C167.496,28.456,130.766,4,85.515,4
		C40.275,4,4,28.456,4,58.619c0,26.291,27.361,48.249,65.361,53.448V112.063z"
              stroke="#FF69B4"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-miterlimit="10"
              stroke-width="14"
            />
            <path
              class="speech-bubble-fill"
              d="M69.361,109.063l16.377,28.99l16.562-28.94
		c37.277-5.125,65.195-27.13,65.195-53.494C167.496,25.456,130.766,1,85.515,1C40.275,1,4,25.456,4,55.619
    c0,26.291,27.361,48.249,65.361,53.448V109.063z"
              fill="#ffffff"
            />
            <text
              class="speech-bubble-text"
              fill="#FF69B4"
              style="text-anchor: middle;"
              x="85"
              y="67"
            />
          </g>
          <g
            transform="matrix(1,0,0,1,404 215)"
          >
            <g
              filter="url(#goo)"
              style="pointer-events: initial; user-select: text; cursor: grab;"
            >
              <rect
                style="fill: rgba(0, 0, 0, 0); height: 100%;"
                transform="matrix(1,0,0,1,-20,-64)"
                width="808"
              />
              <circle
                class="reaction-dot"
                cx="0"
                cy="16"
                fill="#FF69B4"
                id="dot-happy-0"
                r="10"
                style="cursor: move;"
              />
              <circle
                class="reaction-dot"
                cx="60"
                cy="16"
                fill="#FF69B4"
                id="dot-cool-1"
                r="10"
                style="cursor: pointer;"
              />
              <circle
                class="reaction-dot"
                cx="120"
                cy="16"
                fill="#FF69B4"
                id="dot-evil-2"
                r="10"
                style="cursor: pointer;"
              />
              <circle
                class="reaction-dot"
                cx="180"
                cy="16"
                fill="#FF69B4"
                id="dot-confused-3"
                r="10"
                style="cursor: pointer;"
              />
              <circle
                class="reaction-dot"
                cx="240"
                cy="16"
                fill="#FF69B4"
                id="dot-sad-4"
                r="10"
                style="cursor: pointer;"
              />
              <circle
                class="reaction-dot"
                cx="300"
                cy="16"
                fill="#FF69B4"
                id="dot-shocked-5"
                r="10"
                style="cursor: pointer;"
              />
              <circle
                class="reaction-dot"
                cx="360"
                cy="16"
                fill="#FF69B4"
                id="dot-angry-6"
                r="10"
                style="cursor: pointer;"
              />
              <circle
                class="reaction-dot"
                cx="420"
                cy="16"
                fill="#FF69B4"
                id="dot-wondering-7"
                r="10"
                style="cursor: pointer;"
              />
              <circle
                class="reaction-dot"
                cx="480"
                cy="16"
                fill="#FF69B4"
                id="dot-baffled-8"
                r="10"
                style="cursor: pointer;"
              />
              <circle
                class="reaction-dot"
                cx="540"
                cy="16"
                fill="#FF69B4"
                id="dot-smile-9"
                r="10"
                style="cursor: pointer;"
              />
              <circle
                class="reaction-dot"
                cx="600"
                cy="16"
                fill="#FF69B4"
                id="dot-sleepy-10"
                r="10"
                style="cursor: pointer;"
              />
              <circle
                class="reaction-dot"
                cx="660"
                cy="16"
                fill="#FF69B4"
                id="dot-tongue-11"
                r="10"
                style="cursor: pointer;"
              />
              <circle
                class="reaction-dot"
                cx="720"
                cy="16"
                fill="#FF69B4"
                id="dot-grin-12"
                r="10"
                style="cursor: pointer;"
              />
              <circle
                class="reaction-dot"
                cx="780"
                cy="16"
                fill="#FF69B4"
                id="dot-neutral-13"
                r="10"
                style="cursor: pointer;"
              />
            </g>
            <g>
              <path
                class="reaction-icon"
                d="M16,0C7.2,0,0,7.2,0,16s7.2,16,16,16s16-7.2,16-16S24.8,0,16,0z M22,8c1.1,0,2,1.3,2,3s-0.9,3-2,3s-2-1.3-2-3S20.9,8,22,8z M10,8c1.1,0,2,1.3,2,3s-0.9,3-2,3s-2-1.3-2-3S8.9,8,10,8z M16,28c-5.2,0-9.5-4.4-10-9.9c2.9,1.7,6.4,2.7,10,2.7s7.1-1,10-2.7C25.5,23.6,21.2,28,16,28L16,28z"
                data-index="0"
                fill="#ffffff"
                id="icon-happy-0"
                opacity="0"
                style="pointer-events: none; transform-origin: 50% 50%; transform: scale(0, 0) ;"
                transform="matrix(1,0,0,1,-16,0)"
              />
              <path
                class="reaction-icon"
                d="M16,0C7.2,0,0,7.2,0,16s7.2,16,16,16s16-7.2,16-16S24.8,0,16,0z M16,26c-1.5,0-2.9-0.3-4.2-0.9l1-1.7c1,0.4,2.1,0.7,3.2,0.7c2.9,0,5.5-1.6,6.9-3.9l1.7,1C22.8,24.1,19.6,26,16,26L16,26z M26,12c0,1.1-0.9,2-2,2h-4c-1.1,0-2-0.9-2-2h-4c0,1.1-0.9,2-2,2H8c-1.1,0-2-0.9-2-2V9c0-0.6,0.4-1,1-1h6c0.6,0,1,0.4,1,1v1h4V9c0-0.6,0.5-1,1-1h6c0.5,0,1,0.4,1,1V12z"
                data-index="1"
                fill="#ffffff"
                id="icon-cool-1"
                opacity="0"
                style="pointer-events: none; transform-origin: 50% 50%; transform: scale(0, 0) ;"
                transform="matrix(1,0,0,1,44,0)"
              />
              <path
                class="reaction-icon"
                d="M32,2c0-1.4-0.3-2.8-0.8-4c-1,2.4-3,4.3-5.5,5.3C23,1.2,19.7,0,16,0S9,1.2,6.3,3.3C3.8,2.3,1.9,0.4,0.8-2C0.3-0.8,0,0.6,0,2c0,2.3,0.8,4.4,2.1,6.1C0.8,10.4,0,13.1,0,16c0,8.8,7.2,16,16,16s16-7.2,16-16c0-2.9-0.8-5.6-2.1-7.9C31.2,6.4,32,4.3,32,2z M18,11.9c0.1-1.5,1.4-2.5,2.5-3C21.6,8.3,22.7,8,22.8,8c0.5-0.1,1.1,0.2,1.2,0.7s-0.2,1.1-0.7,1.2c-0.6,0.1-1.2,0.4-1.8,0.7C21.8,11,22,11.5,22,12c0,1.1-0.9,2-2,2s-2-0.9-2-2C18,12,18,11.9,18,11.9L18,11.9z M8,8.8C8.2,8.2,8.7,7.9,9.2,8c0,0,1.1,0.3,2.2,0.8c1.1,0.6,2.5,1.6,2.6,3c0,0,0,0.1,0,0.1c0,1.1-0.9,2-2,2s-2-0.9-2-2c0-0.5,0.2-1,0.5-1.4C10,10.4,9.3,10.1,8.8,10C8.2,9.8,7.9,9.3,8,8.8L8,8.8z M16,26c-3.6,0-6.8-1.9-8.6-4.9l2.6-1.5c1.2,2,3.5,3.4,6,3.4s4.8-1.4,6-3.4l2.6,1.5C22.8,24.1,19.6,26,16,26z"
                data-index="2"
                fill="#ffffff"
                id="icon-evil-2"
                opacity="0"
                style="pointer-events: none; transform-origin: 50% 50%; transform: scale(0, 0) ;"
                transform="matrix(1,0,0,1,104,0)"
              />
              <path
                class="reaction-icon"
                d="M16,0C7.2,0,0,7.2,0,16s7.2,16,16,16s16-7.2,16-16S24.8,0,16,0z M22,8c1.1,0,2,0.9,2,2s-0.9,2-2,2s-2-0.9-2-2S20.9,8,22,8z M10,8c1.1,0,2,0.9,2,2s-0.9,2-2,2s-2-0.9-2-2S8.9,8,10,8z M21.5,25.3c-2.6,0.9-5.5-0.4-6.4-3c-0.6-1.6-2.3-2.4-3.8-1.8c-1.4,0.5-2.2,2-1.9,3.5h-2c-0.3-2.3,1-4.5,3.2-5.3c2.6-0.9,5.5,0.4,6.4,3c0.6,1.6,2.3,2.4,3.8,1.8c1.4-0.5,2.2-2,1.9-3.5h2C25,22.3,23.7,24.5,21.5,25.3z"
                data-index="3"
                fill="#ffffff"
                id="icon-confused-3"
                opacity="0"
                style="pointer-events: none; transform-origin: 50% 50%; transform: scale(0, 0) ;"
                transform="matrix(1,0,0,1,164,0)"
              />
              <path
                class="reaction-icon"
                d="M16,0C7.2,0,0,7.2,0,16s7.2,16,16,16s16-7.2,16-16S24.8,0,16,0z M22,8c1.1,0,2,0.9,2,2s-0.9,2-2,2s-2-0.9-2-2S20.9,8,22,8z M10,8c1.1,0,2,0.9,2,2s-0.9,2-2,2s-2-0.9-2-2S8.9,8,10,8z M22,24.4c-1.2-2-3.5-3.4-6-3.4s-4.8,1.4-6,3.4l-2.6-1.5C9.2,19.9,12.4,18,16,18s6.8,1.9,8.6,4.9L22,24.4z"
                data-index="4"
                fill="#ffffff"
                id="icon-sad-4"
                opacity="0"
                style="pointer-events: none; transform-origin: 50% 50%; transform: scale(0, 0) ;"
                transform="matrix(1,0,0,1,224,0)"
              />
              <path
                class="reaction-icon"
                d="M16,0C7.2,0,0,7.2,0,16s7.2,16,16,16s16-7.2,16-16S24.8,0,16,0z M10,14c-1.1,0-2-1.3-2-3s0.9-3,2-3s2,1.3,2,3S11.1,14,10,14z M16,26c-2.2,0-4-1.8-4-4s1.8-4,4-4s4,1.8,4,4S18.2,26,16,26z M22,14c-1.1,0-2-1.3-2-3s0.9-3,2-3s2,1.3,2,3S23.1,14,22,14z"
                data-index="5"
                fill="#ffffff"
                id="icon-shocked-5"
                opacity="0"
                style="pointer-events: none; transform-origin: 50% 50%; transform: scale(0, 0) ;"
                transform="matrix(1,0,0,1,284,0)"
              />
              <path
                class="reaction-icon"
                d="M16,0C7.2,0,0,7.2,0,16s7.2,16,16,16s16-7.2,16-16S24.8,0,16,0z M18,11.9c0.1-1.5,1.4-2.5,2.5-3C21.6,8.3,22.7,8,22.8,8c0.5-0.1,1.1,0.2,1.2,0.7s-0.2,1.1-0.7,1.2c-0.6,0.1-1.2,0.4-1.8,0.7C21.8,11,22,11.5,22,12c0,1.1-0.9,2-2,2s-2-0.9-2-2C18,12,18,11.9,18,11.9L18,11.9z M8,8.8C8.2,8.2,8.7,7.9,9.2,8c0,0,1.1,0.3,2.2,0.8c1.1,0.6,2.5,1.6,2.6,3c0,0,0,0.1,0,0.1c0,1.1-0.9,2-2,2s-2-0.9-2-2c0-0.5,0.2-1,0.5-1.4C10,10.4,9.3,10.1,8.8,10C8.2,9.8,7.9,9.3,8,8.8L8,8.8z M22,24.4c-1.2-2-3.5-3.4-6-3.4s-4.8,1.4-6,3.4l-2.6-1.5C9.2,19.9,12.4,18,16,18s6.8,1.9,8.6,4.9L22,24.4z"
                data-index="6"
                fill="#ffffff"
                id="icon-angry-6"
                opacity="0"
                style="pointer-events: none; transform-origin: 50% 50%; transform: scale(0, 0) ;"
                transform="matrix(1,0,0,1,344,0)"
              />
              <path
                class="reaction-icon"
                d="M16,0C7.2,0,0,7.2,0,16s7.2,16,16,16s16-7.2,16-16S24.8,0,16,0z M22,8c1.1,0,2,0.9,2,2s-0.9,2-2,2s-2-0.9-2-2S20.9,8,22,8z M8,10c0-1.1,0.9-2,2-2s2,0.9,2,2s-0.9,2-2,2S8,11.1,8,10z M10.4,25.2l-0.7-2.4l13.7-4l0.7,2.4L10.4,25.2z"
                data-index="7"
                fill="#ffffff"
                id="icon-wondering-7"
                opacity="0"
                style="pointer-events: none; transform-origin: 50% 50%; transform: scale(0, 0) ;"
                transform="matrix(1,0,0,1,404,0)"
              />
              <path
                class="reaction-icon"
                d="M16,0C7.2,0,0,7.2,0,16s7.2,16,16,16s16-7.2,16-16S24.8,0,16,0z M8,13c0-1.7,1.3-3,3-3s3,1.3,3,3s-1.3,3-3,3S8,14.7,8,13z M20,24h-8v-2h8V24z M21,16c-1.7,0-3-1.3-3-3s1.3-3,3-3s3,1.3,3,3S22.7,16,21,16z M12,13c0,0.6-0.4,1-1,1s-1-0.4-1-1s0.4-1,1-1S12,12.4,12,13z M22,13c0,0.6-0.4,1-1,1s-1-0.4-1-1s0.4-1,1-1S22,12.4,22,13z"
                data-index="8"
                fill="#ffffff"
                id="icon-baffled-8"
                opacity="0"
                style="pointer-events: none; transform-origin: 50% 50%; transform: scale(0, 0) ;"
                transform="matrix(1,0,0,1,464,0)"
              />
              <path
                class="reaction-icon"
                d="M16,0C7.2,0,0,7.2,0,16s7.2,16,16,16s16-7.2,16-16S24.8,0,16,0z M22,8c1.1,0,2,0.9,2,2s-0.9,2-2,2s-2-0.9-2-2S20.9,8,22,8z M10,8c1.1,0,2,0.9,2,2s-0.9,2-2,2s-2-0.9-2-2S8.9,8,10,8z M16,26c-3.6,0-6.8-1.9-8.6-4.9l2.6-1.5c1.2,2,3.5,3.4,6,3.4s4.8-1.4,6-3.4l2.6,1.5C22.8,24.1,19.6,26,16,26z"
                data-index="9"
                fill="#ffffff"
                id="icon-smile-9"
                opacity="0"
                style="pointer-events: none; transform-origin: 50% 50%; transform: scale(0, 0) ;"
                transform="matrix(1,0,0,1,524,0)"
              />
              <path
                class="reaction-icon"
                d="M16,0C7.2,0,0,7.2,0,16s7.2,16,16,16s16-7.2,16-16S24.8,0,16,0z M9.7,10.3c-0.4,0.4-1,0.4-1.4,0s-0.4-1,0-1.4c1.4-1.4,4-1.4,5.4,0c0.4,0.4,0.4,1,0,1.4c-0.2,0.2-0.5,0.3-0.7,0.3s-0.5-0.1-0.7-0.3C11.7,9.7,10.3,9.7,9.7,10.3z M16,26c-2.2,0-4-2.2-4-5s1.8-5,4-5s4,2.2,4,5S18.2,26,16,26z M23.7,10.3c-0.2,0.2-0.5,0.3-0.7,0.3s-0.5-0.1-0.7-0.3c-0.6-0.6-2-0.6-2.6,0c-0.4,0.4-1,0.4-1.4,0s-0.4-1,0-1.4c1.4-1.4,4-1.4,5.4,0C24.1,9.3,24.1,9.9,23.7,10.3z"
                data-index="10"
                fill="#ffffff"
                id="icon-sleepy-10"
                opacity="0"
                style="pointer-events: none; transform-origin: 50% 50%; transform: scale(0, 0) ;"
                transform="matrix(1,0,0,1,584,0)"
              />
              <path
                class="reaction-icon"
                d="M16,0C7.2,0,0,7.2,0,16s7.2,16,16,16s16-7.2,16-16S24.8,0,16,0z M10,8c1.1,0,2,0.9,2,2s-0.9,2-2,2s-2-0.9-2-2S8.9,8,10,8z M24,20h-2v3c0,1.7-1.3,3-3,3s-3-1.3-3-3v-3H8v-2h16V20z M22,12c-1.1,0-2-0.9-2-2s0.9-2,2-2s2,0.9,2,2S23.1,12,22,12z"
                data-index="11"
                fill="#ffffff"
                id="icon-tongue-11"
                opacity="0"
                style="pointer-events: none; transform-origin: 50% 50%; transform: scale(0, 0) ;"
                transform="matrix(1,0,0,1,644,0)"
              />
              <path
                class="reaction-icon"
                d="M16,0C7.2,0,0,7.2,0,16s7.2,16,16,16s16-7.2,16-16S24.8,0,16,0z M22,7.4c2,0,3.6,1.6,3.6,3.6c0,0.2,0,0.4-0.1,0.6c-0.1,0.3-0.3,0.5-0.6,0.5s-0.6-0.2-0.6-0.5c-0.2-1.1-1.2-1.7-2.3-1.7s-2.1,0.5-2.3,1.7c0,0.3-0.3,0.5-0.6,0.5l0,0c-0.3,0-0.6-0.2-0.6-0.5c0-0.2-0.1-0.4-0.1-0.6C18.4,9,20,7.4,22,7.4L22,7.4z M10,7.4c2,0,3.6,1.6,3.6,3.6c0,0.2,0,0.4-0.1,0.6c-0.1,0.3-0.3,0.5-0.6,0.5s-0.6-0.2-0.6-0.5c-0.2-1.1-1.2-1.7-2.3-1.7s-2.1,0.5-2.3,1.7c-0.1,0.3-0.3,0.5-0.6,0.5l0,0c-0.3,0-0.6-0.2-0.6-0.5c0-0.2-0.1-0.4-0.1-0.6C6.4,9,8,7.4,10,7.4L10,7.4z M6,18h6v7.7C8.6,24.9,6,21.7,6,18z M14,26v-8h4v8H14zM20,25.7V18h6C26,21.7,23.4,24.9,20,25.7z"
                data-index="12"
                fill="#ffffff"
                id="icon-grin-12"
                opacity="0"
                style="pointer-events: none; transform-origin: 50% 50%; transform: scale(0, 0) ;"
                transform="matrix(1,0,0,1,704,0)"
              />
              <path
                class="reaction-icon"
                d="M16,0C7.2,0,0,7.2,0,16s7.2,16,16,16s16-7.2,16-16S24.8,0,16,0z M20,24h-8v-2h8V24z M22,8c1.1,0,2,0.9,2,2s-0.9,2-2,2s-2-0.9-2-2S20.9,8,22,8z M10,8c1.1,0,2,0.9,2,2s-0.9,2-2,2s-2-0.9-2-2S8.9,8,10,8z"
                data-index="13"
                fill="#ffffff"
                id="icon-neutral-13"
                opacity="0"
                style="pointer-events: none; transform-origin: 50% 50%; transform: scale(0, 0) ;"
                transform="matrix(1,0,0,1,764,0)"
              />
            </g>
          </g>
        </svg>
      </div>
    </div>
  </div>
  <div
    class="svg-bubble-action"
  />
</DocumentFragment>
`;
